### ISSUE TITLE
Wildcard Domains (not SSL)

### ISSUE THREAD

**Issue by markwylde** [jc21/nginx-proxy-manager#749]
Great project, thanks for all your time on this.

This feature may already exist, and I just can't find any documentation for it. I'm happy to create a PR with docs if you can point me in the right direction.

Basically, everything is working perfectly with my new installation. However, I have to put each subdomain in manually. I would like to specify a wildcard domain, and have every subdomain map to a host.

Just to note, this isn't to do with HTTPS or LetsEncrypt.

**Is your feature request related to a problem? Please describe.**
I would like to be able to map `*.mydomain.com` to the same IP address.

**Describe the solution you'd like**
The screen below to accept a wildcard domain
![image](https://user-images.githubusercontent.com/5929807/100749548-85098400-3430-11eb-8c01-74eafed92737.png)

**Describe alternatives you've considered**
Putting in each domain manually

**Additional context**

**Use Case**
Maybe a use case would help. Imagine I'm hosting GitLab (or a hypothetical GitHub) on my network with pages (GitHub Pages or GitLab Pages).

I want to let my users expose `THEIRORG.pages.gitlab.test` .

In my mind, the way to do this is have a wildcard entry: `*.pages.gitlab.test` 

So then if the user comes through to `http://testorg.pages.gitlab.test/testrepo`, the request automaitcally get's forwarded to the one server.

**Comment by burkeazbill** [comment]
Looking for this as well. I can do with Traefik, but want to use NPM for others to more easily do things via UI rather than config files ;)

**Comment by 1xPdd** [comment]
I've had success essentially using 2 certificates: `*.example.com` for all subdomains and a separate certificate for `example.com`.

**Comment by markwylde** [comment]
> I've had success essentially using 2 certificates: `*.example.com` for all subdomains and a separate certificate for `example.com`.

Just to clarify, this ticket isn't in regards to wildcard SSL certificates. But wildcard domain name matching. Essentially, you could load balance to two http (not https) endpoints. IE:

For the following:
http://test1.example.com -> 192.168.1.10
http://test2.example.com -> 192.168.1.10
http://anything.example.com -> 192.168.1.10

The rule would be:
http://*.example.com -> 192.168.1.10

**Comment by hhannis** [comment]
yes, want this!

**Comment by mosaati** [comment]
AFAIK, to be able to ask nginx to route a specific domain to a specific IP address it has to know to which port. Having all domain and subdomains routed to the same IP address wouldn't make sense because they would require different ports for them to function.

You could have them all on the same IP, that doesn't matter much. But let's take your examples http://test1.example.com -> 192.168.1.10 and http://test2.example.com -> 192.168.1.10 are for example running on ports 8080 and 8081 (they cannot have the same port since they are served on the same IP) you would have to add 2 entries in NPM to define those ports separately.

If you mean strictly load balancing, it could be done by having a stream configuration. Although it is an area that might require some further work.

**Comment by markwylde** [comment]
Hey @mosaati, thanks for your comment.

That's exactly the scenario I want. I wnat both test1 and test2 to point to 192.168.1.10:80 (the same port).

Take an example nodejs server:

```javascript
const http = require('http');

function handler (request, response) {
  if (request.headers.host === 'test1.example.com') {
    response.end('Hi, I am test1');
    return;
  }

  if (request.headers.host === 'test2.example.com') {
    response.end('Hi, I am test2');
    return;
  }

  response.end('There is no website hosted at ' + request.headers.host);
}

const server = http.createServer(handler);

server.listen(8000);
````

So to clarify: I would like to point a wildcard domain to the same IP **and port**.

**Comment by mosaati** [comment]
Then I think you would need a stream configuration. Have you tried it?

**Comment by schniper** [comment]
> AFAIK, to be able to ask nginx to route a specific domain to a specific IP address it has to know to which port. Having all domain and subdomains routed to the same IP address wouldn't make sense because they would require different ports for them to function.
> 
> You could have them all on the same IP, that doesn't matter much. But let's take your examples http://test1.example.com -> 192.168.1.10 and http://test2.example.com -> 192.168.1.10 are for example running on ports 8080 and 8081 (they cannot have the same port since they are served on the same IP) you would have to add 2 entries in NPM to define those ports separately.
> 
> If you mean strictly load balancing, it could be done by having a stream configuration. Although it is an area that might require some further work.

This is not accurate. I am currently using the following config to serve websites from all subfolders in a main folder, using something like this:

server {
    listen 80 default_server;
    server_name ~^(?<sname>.+)\.loc$ _;
    
    include nginx-generic;
}
server {
    listen 443 default_server;
    server_name ~^(?<sname>.+)\.loc$ _;
    
    ssl_certificate /etc/nginx/nginx.crt;
    ssl_certificate_key /etc/nginx/nginx.key;
    include nginx-generic-ssl;
}

And then reference the active domain using the $sname variable. So routing based on domain name. All on the same port:

root "/var/www/$sname/public";

location / {
  root  "/var/www/$sname/public";
  try_files $uri $uri/ /index.php$is_args$args;
  autoindex off;
  index index.html index.htm index.php;
}

etc...

This would indeed be a very useful feature.
The regular expression should be editable, so that, for example, we could have a separate route for the admin site on the same domain or different route sets for other specific websites (I am using exclusions to leave out websites which need specific rules, like cms ones with special configs, etc.)

The idea is this is totally doable.

**Comment by hydridity** [comment]
Indeed would be nice to have support for this, i'm using ngix-proxy-manager as entry point behind NAT for multiple services being accessible via different subdomains pointed to one public ip.

Today i needed to test system that would serve additional resources based on consuming the header containing wildcard subdomain of subdomain (for example *.test.domain.com) and because the nature of this is dynamic, i cannot just set static proxy hosts in ngix-proxy manager because i can never know what the next one will be.
I think the same problem would occur when two reverse proxy layers are required, while the first one would simply redirect all requests of domain or subdomain and all of it's wildcard subdomains, the second one would then do all the required redirection.

Right now it is not possible when the ngix-proxy-manager is the first reverse proxy, or i have missed something important.

**Comment by mantasio** [comment]
This would be a really useful feature. Especially if you are working in development with CI/CD and want to have different environments for, let's say, each git branch. Right now, you would need to add a new entry in NPM each time after deployment is done, which isn't very handy

**Comment by PraUrb** [comment]
I'm not quite sure whether that's exactly what you want, but maybe you could try this:

Create a new Proxy Host with the name "example.org" and assign it the SSL Certificate for *.example.com
In the "Advanced" tab add this: `server_name  *.example.org;`

If you add a new host now, e.g. "test.example.org", Nginx Proxy Manager will pass the request to the IP defined on this specific host, because its server name matches exactly and it therefor has a higher priority. But if you try to access "test2.example.org" (which you haven't explicitly defined!), it will pass the request to the Proxy Host with the name "example.org".

**Comment by markwylde** [comment]
It might be close, but I think the problem is you'd still have to add those hosts manually.

Maybe a use case would help. Imagine I'm hosting GitLab (or a hypothetical GitHub) on my network with pages (GitHub Pages or GitLab Pages).

I want to let my users expose `THEIRORG.pages.gitlab.test` .

In my mind, the way to do this is have a wildcard entry: `*.pages.gitlab.test` 

So then if the user comes through to `http://testorg.pages.gitlab.test/testrepo`, the request automaitcally get's forwarded to the one server.

**Comment by PFindersDomain** [comment]
How does one enter *.domain.com as this does not show the add option when I type this in, but this screenshot shows it added.
![proxy-hosts-add](https://user-images.githubusercontent.com/14099213/127672547-3c5a15f4-e716-472b-bee2-a28d20705af0.png)

**Comment by kyoumei** [comment]
Is it possible to extend support for this to sub subdomains? ie. *.test.domain.com ?

**Comment by chaptergy** [comment]
It seems the only issue was the reges pattern. The fix was added in v2.9.5, so from there and up this works. I have tested it on v2.9.6 and both first-level and second-level wildcards for proxy hosts work as expected.
